### DISCUSSION TITLE
: fvalue [inlined] 2value ;

### DISCUSSION THREAD

**Discussion by robzed** [tabemann/zeptoforth#66] [Q&A]
When I'm compile-to-flash, fvalue's don't appear to have the correct value. And resetting doesn't show the correct value?

compile-to-flash
#include float_compatibility.fth
#include config.fth

Files can be found here: https://github.com/robzed/ukmarsbot_forth/tree/master/for_rp2040

**Comment by tabemann** [comment]
I tried reproducing this and was unable to. However, I should warn you that many fixed point numbers with digits to the right of the decimal point will not look "correct" due to not having any exact binary value.

**Reply by robzed** [reply]
Thanks for trying. I will try again tomorrow and post my results.

**Reply by tabemann** [reply]
Oh hey I think I figured out your issue! The matter is likely that when you compile a `value` or `2value` in flash compilation mode you cannot read their value without rebooting first because it will get garbage (because it will be trying to read the address that is alloted on boot for the `value` or `2value` which will be used by something else or by nothing in particular first). Have you tried rebooting before trying to use a `value` or `2value`'s value?

**Reply by tabemann** [reply]
This is why I should read comments more closely - when I ran your code under zeptoforth, I had forgotten to add `compile-to-flash` beforehand (and when I did compile to flash when doing cursory tests with `value` and `2value` I had reflexively rebooted the MCU before utilizing the `value`s I had just created).

**Reply by robzed** [reply]
As you know, getting issue reports is half figuring what the tester/user means and half where the problem is! Fixing often is often easy after these two.

I think this is a big gotchya. If we can't fix this somehow in code then we should mention this next to the value and 2value descriptions and on the wiki in the section about Value.

**Reply by robzed** [reply]
By the way I found value/2value ideal for the use case here. They are effectively modifiable constants - which really works well for this configuration variables. 

True - extra work would be required to make modified values persistent - but I think this is good.  I don't want to save temporary test values every time I try something and lose the original value. 

So definitely worth the effort it too you to add values - thanks.

**Reply by robzed** [reply]
I think I've figured out the problem with config.fth. Even after a reboot the value isn't correct - because they are calculated **from other values**

I figured this out by adding a print before defining: 
`calc_deg/mm_diff  fdup f.  fvalue DEG/mm_DIFFERENCE 0,0000003350  ok`

It should be 0,8952465557. 

To repeat - of course the value we are storing isn't correct here ... because it's calculated based on other values, which are not correct until reboot. I'll have a think about how to get around this - but this use of value/2value isn't going to work here as it is. 

### Extra info

Using the following:
```
compile-to-flash
#include float_compatibility.fth
#include config.fth
```

For reference below are the results I get.

### from end of config.fth

: calc_deg/mm_diff ( -- f ) 180,0 2,0 f/ WHEEL_CIRCUM f/ ;  ok
calc_deg/mm_diff    fvalue DEG/mm_DIFFERENCE  ok
: calc-config  ok
calc_wheel_circum   to WHEEL_CIRCUM  \ mm  ok
calc_pulses/rev     to PULSES/REV  ok
calc_mm/count       to BASIC_mm/count  ok
calc_mm/count_left  to mm/COUNT_LEFT  ok
calc_mm/count_right to mm/COUNT_RIGHT  ok
calc_deg/mm_diff    to DEG/mm_DIFFERENCE  ok
;  ok
: show-config  ok
." mm/COUNT_LEFT " mm/COUNT_LEFT f. cr  ok
." mm/COUNT_RIGHT " mm/COUNT_RIGHT f. cr  ok
." DEG/mm_DIFFERENCE " DEG/mm_DIFFERENCE f. cr  ok
;  ok
DEG/mm_DIFFERENCE .s  [ 1573941093 1917799667 ] ok
****f. 1917799667,3664617177  ok**
**reboot**
Welcome to zeptoforth
Built for rp2040, version 1.1.0, on Mon Sep 11 07:21:28 PM CDT 2023
zeptoforth comes with ABSOLUTELY NO WARRANTY: for details type `license'
 ok
**DEG/mm_DIFFERENCE .s  [ 1439 0 ] ok
f. 0,0000003350  ok**
calc-config  ok
**DEG/mm_DIFFERENCE .s  [ -449912617 0 ] ok
f. 0,8952465557  ok**

**Reply by tabemann** [reply]
What I would probably do is to create the initial values as `constant`s so they can be utilized when compiling to flash and use the generated values as the default values for the `value`s in such a fashion that the `value`s are never read until after rebooting.

**Reply by robzed** [reply]
That's a good idea using constants as defaults.

**Comment by tabemann** [comment]
One other note - there is no need to make `fconstant`, `fvariable`, or `fvalue` inlined because in 99% of use cases these will be executed in interpretation mode.

**Reply by robzed** [reply]
Agreed.

**Reply by robzed** [reply]
Fixed

**Comment by tabemann** [comment]
@robzed Shall I close this issue, or is there anything else here? Thanks for bringing this up, BTW, as you probably have noticed I sometimes forget that other people don't know all the ins and outs of zeptoforth.

**Reply by robzed** [reply]
Fine to close
